File: Leetcode/Searching/33.Search-in-rotated-sorted-array.java
/*
Suppose an array sorted in ascending order is rotated at some pivot unknown to you beforehand.

(i.e., [0,1,2,4,5,6,7] might become [4,5,6,7,0,1,2]).

You are given a target value to search. If found in the array return its index, otherwise return -1.

You may assume no duplicate exists in the array.

Your algorithm's runtime complexity must be in the order of O(log n).

Example 1:

Input: nums = [4,5,6,7,0,1,2], target = 0
Output: 4

Example 2:

Input: nums = [4,5,6,7,0,1,2], target = 3
Output: -1

********************************************************************Solution*****************************************************************************/

class Solution 
{
    public int search(int[] a, int target) 
    {
        if(a.length == 0)
            return -1;

        int start = 0, end = a.length - 1;

        //in this case array is rotated
        if(a[start] > a[end])
        {            
            int indexPivot = findPivot(a, start, end, a.length);

            if(target >= a[start])
                return binarySearch(a, start, indexPivot - 1, target);

            return binarySearch(a, indexPivot, end, target);
        }   

        //if array is sorted but not rotated
        return binarySearch(a, start, end, target);
     } 


    //this will return the index of minimum or pivot element of the rotated sorted array
    int findPivot(int a[], int start, int end, int size)
    {
        int toCompare = a[start];
        
        while(start <= end)
        {
            int mid = start + ((end - start) / 2);
            
            int prevMid = (mid + size - 1) % size;
            int nextMid = (mid + 1) % size;

            if(a[mid] < a[prevMid] && a[mid] < a[nextMid])
                return mid;

            else if(a[mid] >= toCompare)
                start = mid + 1;

            else if(a[mid] < toCompare)
                end = mid - 1;
        }  

        return -1;
    } 

    
    //this will return the index of target 
    int binarySearch(int a[], int start, int end, int target)
    {
        while(start <= end)
        {
            int mid = start + ((end - start) / 2);

            if(a[mid] == target)
                return mid;

            else if(a[mid] < target)
                start = mid + 1;

            else //if(a[mid] > target)
                end = mid - 1;
        }

        return -1;
    } 
    
}

File: DSA-JavaScript/Searching/33.search-in-rotated-sorted-array.js
/*

There is an integer array nums sorted in ascending order (with distinct values).

Prior to being passed to your function, nums is possibly rotated at an unknown pivot index k (1 <= k < nums.length) such that the resulting array is [nums[k], nums[k+1], ..., nums[n-1], nums[0], nums[1], ..., nums[k-1]] (0-indexed). For example, [0,1,2,4,5,6,7] might be rotated at pivot index 3 and become [4,5,6,7,0,1,2].

Given the array nums after the possible rotation and an integer target, return the index of target if it is in nums, or -1 if it is not in nums.

You must write an algorithm with O(log n) runtime complexity.

 
Example 1:

Input: nums = [4,5,6,7,0,1,2], target = 0
Output: 4

Example 2:

Input: nums = [4,5,6,7,0,1,2], target = 3
Output: -1

Example 3:

Input: nums = [1], target = 0
Output: -1
 

Constraints:

1 <= nums.length <= 5000
-10^4 <= nums[i] <= 10^4
All values of nums are unique.
nums is an ascending array that is possibly rotated.
-10^4 <= target <= 10^4

***************************************************************Solution*******************************************************/

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
const search = function (nums, target) {
  let low = 0,
    high = nums.length - 1,
    mid;

  while (low <= high) {
    mid = low + Math.floor((high - low) / 2);

    if (nums[mid] === target) {
      return mid;
    }

    /* Comparing mid item with the 1st item in the list to know whether the left half is sorted or not, if the mid item 
    is greater than or equal to the 1st item in the list then we know that we're in the left half of the array which is sorted. */
    if (nums[low] <= nums[mid]) {
      /* Next, we need to check whether the target item lies in the left half or not(1stItem <=target < midItem), 
      if it lies in that range, update the high pointer to search in the left half part, else we need to update 
      the low pointer to discard the left half part */
      if (target >= nums[low] && target < nums[mid]) {
        high = mid - 1;
      } else {
        low = mid + 1;
      }
    } else {
      if (target > nums[mid] && target <= nums[high]) {
        low = mid + 1;
      } else {
        high = mid - 1;
      }
    }
  }

  return -1;
};
